src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.cs

```diff
@@ -2,33 +2,74 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+
 namespace System.Threading
 {
     /// <summary>
     /// A LIFO semaphore.
     /// Waits on this semaphore are uninterruptible.
     /// </summary>
-    internal sealed class LowLevelLifoSemaphore : IDisposable
+    internal sealed partial class LowLevelLifoSemaphore : IDisposable
     {
-        private WaitSubsystem.WaitableObject _semaphore;
+        private WaiterListEntry _waiterStackHead;
+        private LowLevelLock _waiterStackLock;
+        [ThreadStatic]
+        private static WaiterListEntry t_waitEntry;
 
-        public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount)
+        private void Create(int maximumSignalCount)
         {
-            _semaphore = WaitSubsystem.WaitableObject.NewSemaphore(initialSignalCount, maximumSignalCount);
+            _waiterStackHead = null;
+            _waiterStackLock = new LowLevelLock();
         }
 
-        public bool Wait(int timeoutMs)
+        public void Dispose()
         {
-            return WaitSubsystem.Wait(_semaphore, timeoutMs, false, true);
         }
 
-        public int Release(int count)
+        private bool WaitCore(int timeoutMs)
         {
-            return WaitSubsystem.ReleaseSemaphore(_semaphore, count);
+            WaiterListEntry waitEntry = t_waitEntry ?? (t_waitEntry = new WaiterListEntry());
+            waitEntry._monitor.Acquire();
+            try
+            {
+                _waiterStackLock.Acquire();
+                waitEntry._next = _waiterStackHead;
+                _waiterStackHead = waitEntry;
+                _waiterStackLock.Release();
+                return waitEntry._monitor.Wait(timeoutMs);
```

Comment:
This is horribly broken and I didn't notice it until I experimented with stack-allocated mutexes/condition. In the case of timeout the object is not removed from the wait queue, which results in subtly broken behavior.

Reply:
I believe Wait(timeout) can also return true for a spurious wake: https://linux.die.net/man/3/pthread_cond_timedwait. May need to track some state in the entry to ensure that the thread woke up due to a signal.

Reply:
Yep, you are right. These spurious waits are not handled in `LowLevelMonitor` either, but it didn't matter for the callers in current code.

I rewrote the whole "LIFO semaphore" thing in unmanaged code to test how would it behave. It uncovered couple of nasty things like this and I'm doing more testing to cover these cases. The nice property about it is that it uses only one pthread condition for each waiting thread and one pthread mutex. It also allows storing the waiting list entry on stack and thus avoiding any heap allocations.

(I'm not necessarily convinced that unmanaged code is the way to go. I'll revert the CoreRT code to use `WaitSubsystem` to move this PR forward and fix all the other little things you commented on. Thanks for all the feedback!)